### PR TITLE
feat: add container breakpoints

### PR DESCRIPTION
**Pull request by colbr**
Adds a container wrapper which:
- Keeps track of container width using `ResizeObserver`
- Writes breakpoint values, in increments of 50, to an attribute on the wrapper (`<div data-container-breakpoints="0-50-100-150-">`)

To use breakpoints in components, you can check if the breakpoint attribute contains the breakpoint you're targeting, e.g. `[data-container-breakpoints*="-50-"]`.

This wrapper can be used anywhere in the component tree to make components even more reusable, but for now we just have one wrapping the whole app.

This seems to a bit less hacky than most solutions I looked into (Raven Technique, `clamp` + `@keyframe` hacks, etc.), granted it is still a little bit hacky.

_note: this is just a POC_

### CODE REVIEW

File: src/lib/util/DynamicSizeWrapper/DynamicSizeWrapper.tsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+			return;
+		}
+
+		const resizeObserver = new ResizeObserver((event) => {
+			const width = event[0].contentBoxSize[0].inlineSize;
+			const breakpointWidths = Array.from(
+				{ length: Math.ceil(width / 50) },
+				(_, i) => i * 50,
+			);
+			setBreakpoints(breakpointWidths.join('-'));
```
Comment (dalefukami) [inline src/lib/util/DynamicSizeWrapper/DynamicSizeWrapper.tsx:30]: This feels unnecessarily granular to me. How many different sizes does the app need to account for? Most really only have a few. Something like, small/medium/large?

Could we just limit to 3 sizes and rather than use a data-attribute just put a class on the main container? `container--small`

**Review by dalefukami — COMMENTED** [review]
Cool idea. Just to make sure I understand the intent here's what I think is happening:
We want the app itself to be able render differently based on the container it's given to render in versus the browser size (and thus media queries don't work). This is because I might have a wide browser window but have numerous sidebars expanded which limits the actual apps width even though the window is large.

Sound right?

**Comment by colbr** [comment]
> Sound right?

Yep correct! I'll reply to other comments individually, but yes this is the jist of it.

File: src/lib/util/DynamicSizeWrapper/DynamicSizeWrapper.tsx (hunk 1 above)
Comment (colbr) [reply src/lib/util/DynamicSizeWrapper/DynamicSizeWrapper.tsx:30]: The problem I generally have with `small`, `medium`, `large`, is that it's rare for elements to naturally fit into `small`, `medium`, and `large` buckets. 

For example, you might have a row of `Card` elements which look good up until `500px`, at which point you'd want to switch them to a column layout. If you have `small = 300`, `medium = 700`, your only choice is to switch to a column layout at `700px`, which means there's a `200px` gap where the `Card` elements will look unnaturally stretched out when they could have instead stayed arranged as a row. The bigger the gaps between your breakpoint definitions, the more you're going to run into cases where the layout has no choice but to adapt earlier than it needs to. You can add more values, like `mobile-s`, `mobile-m`, etc. but I think at that point you may as well just use increments of `50` to reduce complexity.

tl;dr: imo you get a much nicer and consistent UI if you dial in breakpoints to the width that elements actually break at, rather than rounding breakpoints into buckets, but could just be my pedantic design-brain.

File: src/styles/_breakpoints.scss [hunk 2]
```diff
@@ -0,0 +1,7 @@
+@mixin container-breakpoint($value) {
+    [data-container-breakpoints*="-#{$value}"] & {
+        @content;
+    }
+}
+
+$breakpoint-mobile: 400;
```
[the remark this answers is not in the document]
Comment (colbr) [reply src/styles/_breakpoints.scss:7]: Yeah good point!

**Review by domw30 — DISMISSED** [review]
🚀

File: src/lib/util/DynamicSizeWrapper/DynamicSizeWrapper.tsx (hunk 1 above)
Comment (dalefukami) [reply src/lib/util/DynamicSizeWrapper/DynamicSizeWrapper.tsx:30]: Well, that depends on the system the design team is using. Do you have a design right now that you're matching to that requires you to split on numerous values that aren't nicely predefined? It's not a huge deal but it definitely seems overly complex for any design I've ever had to implement.
